### PR TITLE
switch cmake toolchain: support OUTPUT_NAME property

### LINKED ISSUE #237: The switch cmake toolchain does not support main from static library
The following cmake project fails to build when using the switch cmake toolchain:
**CMakeLists.txt**
```cmake
cmake_minimum_required(VERSION 3.8)
project(main_from_lib)

add_library(staticlib static.c)

add_executable(executable exe.c)
target_link_libraries(executable PRIVATE staticlib)
```
**static.c**
```c
int some_func(void);
int main() {
	return some_func();
}
```
**exe.c**
```c
int some_func(void) {
	return 0;
}
```
Building with switch-cmake with the current `devkitpro/devkita64` docker images gives the following output (and error):
```
/opt/devkitpro/portlibs/switch/bin/aarch64-none-elf-cmake -S ../nxsamp/ -B.  && make
-- The C compiler identification is GNU 11.1.0
-- The CXX compiler identification is GNU 11.1.0
-- Check for working C compiler: /opt/devkitpro/devkitA64/bin/aarch64-none-elf-gcc
-- Check for working C compiler: /opt/devkitpro/devkitA64/bin/aarch64-none-elf-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /opt/devkitpro/devkitA64/bin/aarch64-none-elf-g++
-- Check for working CXX compiler: /opt/devkitpro/devkitA64/bin/aarch64-none-elf-g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /conan/ff
Scanning dependencies of target staticlib
[ 25%] Building C object CMakeFiles/staticlib.dir/static.c.obj
[ 50%] Linking C static library libstaticlib.a
[ 50%] Built target staticlib
Scanning dependencies of target executable
[ 75%] Building C object CMakeFiles/executable.dir/exe.c.obj
[100%] Linking C executable executable
/opt/devkitpro/devkitA64/bin/../lib/gcc/aarch64-none-elf/11.1.0/../../../../aarch64-none-elf/bin/ld: /opt/devkitpro/libnx/lib/libnx.a(switch_crt0.o): in function `_start':
/home/fincs/pacman-packages/libnx/src/libnx-4.1.1/nx/source/runtime/switch_crt0.s:70: undefined reference to `main'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/executable.dir/build.make:85: executable] Error 1
make[1]: *** [CMakeFiles/Makefile2:110: CMakeFiles/executable.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```

This can be fixed by adding `-lnx` twice to the linker flags: once before the user object and libraries and once at the end.

Applying the following patch to `${DEVKITPRO}/cmake/Switch.cmake` fixed the error for me:
```diff
--- /opt/devkitpro/cmake/Switch.cmake	2021-06-14 10:23:20.000000000 +0000
+++ ../Switch.cmake	2021-06-22 15:13:09.613154005 +0000
@@ -12,7 +12,7 @@
 set(CMAKE_CXX_FLAGS_INIT "${NX_COMMON_FLAGS}")
 set(CMAKE_ASM_FLAGS_INIT "${NX_COMMON_FLAGS}")
 
-set(CMAKE_EXE_LINKER_FLAGS_INIT "${NX_ARCH_SETTINGS} ${NX_LIB_DIRS} -fPIE -specs=${DEVKITPRO}/libnx/switch.specs")
+set(CMAKE_EXE_LINKER_FLAGS_INIT "${NX_ARCH_SETTINGS} ${NX_LIB_DIRS} -fPIE -specs=${DEVKITPRO}/libnx/switch.specs -lnx")
 
 set(CMAKE_FIND_ROOT_PATH
 	${CMAKE_FIND_ROOT_PATH}
```

### LINKED ISSUE #238: The switch cmake toolchain does not respect the output name properties of targets
When a cmake project changes the [`OUTPUT_NAME`](https://cmake.org/cmake/help/latest/prop_tgt/OUTPUT_NAME.html) property of a target or sets [`CMAKE_EXECUTABLE_SUFFIX`](https://cmake.org/cmake/help/latest/variable/CMAKE_EXECUTABLE_SUFFIX.html), the `nx_generate_nacp` and `nx_create_nro` functions don't work anymore.

Example the following project:
**CMakeLists.txt**
```cmake
cmake_minimum_required(VERSION 3.8)
project(alt_name)

add_executable(exe exe.c)

# Either one of the following makes this project fail to build
set_property(TARGET exe PROPERTY OUTPUT_NAME "awesome_exe")
set(CMAKE_EXECUTABLE_SUFFIX ".elf")

nx_generate_nacp(awesome.nacp NAME exe)
nx_create_nro(exe NACP awesome.nacp)
```

**exe.c**
```c
int main() {
	return 0;
}
```

Using the `devkitpro/devkita64` docker image, building the project fails with the following error:

```
[ 75%] Generating awesome.nacp
/opt/devkitpro/tools/bin/nacptool --create exe "Unspecified Author" 1.0.0 awesome.nacp
[100%] Generating exe.nro
/opt/devkitpro/tools/bin/elf2nro exe exe.nro --icon=/opt/devkitpro/libnx/default_icon.jpg --nacp=awesome.nacp
Failed to open input!
```

The error occurs because the cmake toolchain assumes the output filename is the same as the target name.

I can fix this by applying the following patch:
```patch
--- /opt/devkitpro/cmake/Switch.cmake	2021-06-14 10:23:20.000000000 +0000
+++ ../Switch.cmake	2021-06-22 16:11:14.049765329 +0000
@@ -81,16 +81,19 @@
 
 	add_custom_command(
 		OUTPUT "${outfile}"
-		COMMAND "${NX_NACPTOOL_EXE}" --create "${NACP_NAME}" "${NACP_AUTHOR}" "${NACP_VERSION}" "${outfile}"
+		COMMAND "${NX_NACPTOOL_EXE}" --create "$<TARGET_FILE:${NACP_NAME}>" "${NACP_AUTHOR}" "${NACP_VERSION}" "${outfile}"
+		DEPENDS ${NACP_NAME}
 		VERBATIM
 	)
 endfunction()
 
-function(nx_create_nro prefix)
+function(nx_create_nro target)
 	cmake_parse_arguments(ELF2NRO "NOICON;NONACP" "ICON;NACP;ROMFS" "" ${ARGN})
 
-	set(ELF2NRO_ARGS ${prefix} "${prefix}.nro")
-	set(ELF2NRO_DEPS ${prefix})
+	get_target_property(TARGET_OUTPUT_NAME ${target} OUTPUT_NAME)
+
+	set(ELF2NRO_ARGS "$<TARGET_FILE:${target}>" "${TARGET_OUTPUT_NAME}.nro")
+	set(ELF2NRO_DEPS ${target})
 
 	if (DEFINED ELF2NRO_ICON AND ELF2NRO_NOICON)
 		message(FATAL_ERROR "nx_create_nro: cannot specify ICON and NOICON at the same time")
@@ -105,7 +108,7 @@
 	endif()
 
 	if (NOT DEFINED ELF2NRO_NACP AND NOT ELF2NRO_NONACP)
-		set(ELF2NRO_NACP "${prefix}.default.nacp")
+		set(ELF2NRO_NACP "${target}.default.nacp")
 		nx_generate_nacp(${ELF2NRO_NACP})
 	endif()
 
@@ -139,15 +142,15 @@
 	endif()
 
 	add_custom_command(
-		OUTPUT "${prefix}.nro"
+		OUTPUT "${TARGET_OUTPUT_NAME}.nro"
 		COMMAND "${NX_ELF2NRO_EXE}" ${ELF2NRO_ARGS}
 		DEPENDS ${ELF2NRO_DEPS}
 		VERBATIM
 	)
 
 	add_custom_target(
-		"${prefix}_nro" ALL
-		DEPENDS "${prefix}.nro"
+		"${target}_nro" ALL
+		DEPENDS "${TARGET_OUTPUT_NAME}.nro"
 	)
 endfunction()
 
```

When building the project above, `awesome_exe.elf`, `awesome_exe.nro` and `awesome.nacp` are generated.

### PR DESCRIPTION
**Pull request by madebr**
FIxes #238 

/cc @WinterMute

### CODE REVIEW

**Comment by fincs** [comment]
We're currently testing a way to fix #237 without having to mess around with the link order, and said fix will be applied to libnx instead (it requires no cmake side changes). Would you mind removing that from this PR? In the future, it'd be appreciated to separate different contributions in different PRs, so that we can independently review, test and merge them without one blocking the others.

File: cmake/switch/Switch.cmake [hunk 1]
```diff
@@ -81,16 +81,22 @@ function(nx_generate_nacp outfile)
...
-function(nx_create_nro prefix)
+function(nx_create_nro target)
 	cmake_parse_arguments(ELF2NRO "NOICON;NONACP" "ICON;NACP;ROMFS" "" ${ARGN})
 
-	set(ELF2NRO_ARGS ${prefix} "${prefix}.nro")
-	set(ELF2NRO_DEPS ${prefix})
+	get_target_property(TARGET_OUTPUT_NAME ${target} OUTPUT_NAME)
+	if(NOT TARGET_OUTPUT_NAME)
+		set(TARGET_OUTPUT_NAME "${target}")
+	endif()
```
Comment (fincs) [inline cmake/switch/Switch.cmake:96]: I'm wondering if it is possible to make this too into a generator expression. Preferably replacing the extension with .nro instead of outright appending when applicable.

Comment (madebr) [reply cmake/switch/Switch.cmake:96]: It's used as `OUTPUT` argument of `add_custom_command`.
When using
```cmake
		OUTPUT "$<TARGET_FILE_NAME:${target}>"
```
instead, the following error is emitted:
```
CMake Error at /conan/Switch.cmake:147 (add_custom_command):
  add_custom_command called with OUTPUT containing a "<".  This character is
  not allowed.
Call Stack (most recent call first):
  CMakeLists.txt:11 (nx_create_nro)


-- Configuring incomplete, errors occurred!

```

Comment (fincs) [reply cmake/switch/Switch.cmake:96]: Ah, so it just doesn't support generator expressions in this context. Kinda sucks, but understandable.

File: cmake/switch/Switch.cmake [hunk 2]
```diff
@@ -81,16 +81,22 @@ function(nx_generate_nacp outfile)
 
 	add_custom_command(
 		OUTPUT "${outfile}"
-		COMMAND "${NX_NACPTOOL_EXE}" --create "${NACP_NAME}" "${NACP_AUTHOR}" "${NACP_VERSION}" "${outfile}"
+		COMMAND "${NX_NACPTOOL_EXE}" --create "$<TARGET_FILE:${NACP_NAME}>" "${NACP_AUTHOR}" "${NACP_VERSION}" "${outfile}"
+		DEPENDS ${NACP_NAME}
```
Comment (fincs) [inline cmake/switch/Switch.cmake:85]: Why was this changed? nx_generate_nacp isn't intended to depend on any target.

Comment (madebr) [reply cmake/switch/Switch.cmake:85]: Oops, this is a leftover of my hacking to fix this + my lack of understanding the nx toolchain.
I'll remove this change.

Comment (madebr) [reply cmake/switch/Switch.cmake]: Oops, this is a leftover of my hacking to fix this + my lack of understanding the nx toolchain.
I'll remove this change.

Comment (madebr) [reply cmake/switch/Switch.cmake]: Removed.

**Review by fincs — APPROVED** [review]

**Comment by fincs** [comment]
Heh, one thing that I overlooked. Could you update the PKGBUILD in that folder too? Bump up the pkgver to 1.1.1 and update the sha256sum of the file you modified.
